src/controller/JCMCommissioner.cpp

```diff
@@ -0,0 +1,430 @@
+ *    You may obtain a copy of the License at
+ *
+ *        http://www.apache.org/licenses/LICENSE-2.0
+ *
+ *    Unless required by applicable law or agreed to in writing, software
+ *    distributed under the License is distributed on an "AS IS" BASIS,
+ *    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *    See the License for the specific language governing permissions and
+ *    limitations under the License.
+ */
+
+#include "JCMCommissioner.h"
+
+#include "JCMTrustVerification.h"
+#include <controller/CommissioningDelegate.h>
+#include <app-common/zap-generated/ids/Attributes.h>
+#include <app-common/zap-generated/ids/Clusters.h>
+#include <app/InteractionModelEngine.h>
+#include <lib/core/CHIPCore.h>
+#include <lib/core/CHIPError.h>
+#include <lib/support/AllocatorUtils.h>
+#include <controller/JCMCommissioner.h>
+#include <credentials/CHIPCert.h>
+
+using namespace ::chip;
+using namespace ::chip::app;
+using namespace chip::app::Clusters;
+
+namespace chip {
+namespace Controller {
+
+/*
+ * JCMDeviceCommissioner public interface and override implementation    
+ */
+CHIP_ERROR JCMDeviceCommissioner::StartJCMTrustVerification(DeviceProxy * proxy)
+{
+    ChipLogProgress(Controller, "JCM: Starting Trust Verification");
+
+    VerifyOrReturnError(proxy != nullptr, CHIP_ERROR_INVALID_ARGUMENT);
+
+    mNextStage = JCMTrustVerificationStage::kStarted;
+    mDeviceProxy = proxy;
+
+    AdvanceTrustVerificationStage(JCMTrustVerificationResult::kSuccess);
+    
+    return CHIP_NO_ERROR;  
+}
+
+void JCMDeviceCommissioner::OnTrustVerificationComplete(JCMTrustVerificationResult result)
+{
+    if (result == JCMTrustVerificationResult::kSuccess)
+    {
+        ChipLogProgress(Controller, "JCM: Administrator Device passed JCM Trust Verification");
+
+        CommissioningStageComplete(CHIP_NO_ERROR);
+    }
+    else
+    {
+        ChipLogError(Controller, "JCM: Failed in verifying 'JCM Trust Verification': err %hu",
+                     static_cast<uint16_t>(result));
+
+        CommissioningDelegate::CommissioningReport report;
+        report.Set<JCMTrustVerificationError>(result);
+
+        CommissioningStageComplete(CHIP_ERROR_INTERNAL, report);
+    }
+}
+
+void JCMDeviceCommissioner::ContinueAfterUserConsent(bool consent)
+{
+    if (consent)
+    {
+        ChipLogProgress(Controller, "JCM: User consent granted");
+        AdvanceTrustVerificationStage(JCMTrustVerificationResult::kSuccess);
+    }
+    else
+    {
+        ChipLogError(Controller, "JCM: User denied consent");
+        AdvanceTrustVerificationStage(JCMTrustVerificationResult::KUserDeniedConsent);
+    }
+}
+
+CHIP_ERROR JCMDeviceCommissioner::ParseAdminFabricIndexAndEndpointId(ReadCommissioningInfo & info)
+{
+    auto attributeCache = info.attributes;
+    
+    CHIP_ERROR err = attributeCache->ForEachAttribute(Clusters::JointFabricAdministrator::Id, [this, &attributeCache](const ConcreteAttributePath & path) {
```

Comment:
I think this will return CHIP_NO_ERROR if the AdministratorFabricIndex is not found. At the end of this function I would recommend checking that mInfo.adminFabricIndex contains a valid value.

Reply:
@doru91 Review the new logic